src/charm.py

```diff
@@ -154,13 +177,87 @@ def _on_ingress_revoked(self, event: IngressPerAppRevokedEvent) -> None:
         if self.unit.is_leader():
             logger.info("This app no longer has ingress")
 
+    def _on_openfga_store_created(self, event: OpenFGAStoreCreateEvent) -> None:
+        """Handle openfga store created event."""
+        if not self._peers:
+            self.unit.status = WaitingStatus("Waiting for peer relation")
+            event.defer()
+            return
+
+        if self._get_openfga_store_info():
+            model_id = self._create_openfga_model()
+            self._set_peer_data(key=self._get_version(), data=dict(openfga_model_id=model_id))
+
+            self._handle_status_update_config(event)
+        else:
+            logger.debug("No openfga store info found, deferring the event")
+            event.defer()
+            return
+
+    def _on_openfga_store_removed(self, event: OpenFGAStoreRemovedEvent) -> None:
+        """Handle openfga store removed event."""
+        logger.info("OpenFGA store was removed")
+        self._pop_peer_data(key=self._get_version())
+        self._handle_status_update_config(event)
+
+    def _on_upgrade_charm(self, event: UpgradeCharmEvent) -> None:
+        """Handle charm upgrade event.
+
+        Create a new model to ensure the migration was run.
+        """
+        if not self._peers:
+            self.unit.status = WaitingStatus("Waiting for peer relation")
+            event.defer()
+            return
+
+        if self._get_openfga_store_info():
+            model_id = self._create_openfga_model()
+            self._set_peer_data(key=self._get_version(), data=dict(openfga_model_id=model_id))
+
+    def _get_openfga_store_info(self) -> Optional[OpenfgaProviderAppData]:
+        openfga_info = self.openfga.get_store_info()
+        if not openfga_info:
+            logger.info("No openfga store info available")
+            return None
+
+        return openfga_info
+
+    def _create_openfga_model(self) -> Optional[str]:
+        openfga_info = self._get_openfga_store_info()
+        cmd = [
+            "identity-platform-admin-ui",
+            "create-fga-model",
+            "--fga-api-url",
+            openfga_info.http_api_url,
+            "--fga-api-token",
+            openfga_info.token,
+            "--fga-store-id",
+            openfga_info.store_id,
+        ]
+        try:
+            process = self._container.exec(cmd)
+            stdout, _ = process.wait_output()
+        except ExecError as err:
+            logger.error(f"Exited with code {err.exit_code}. Stderr: {err.stderr}")
+            return
+
+        out_re = r"Created model:\s*(.+)\s*$"
+        model = re.search(out_re, stdout)
+        if model:
+            return model[1]
```

Comment:
Let's move this to a separate file. Similar to how we have done in Kratos, Hydra and Openfga where we have a class for interacting with the managed application's CLI/API. (e.g. https://github.com/canonical/hydra-operator/blob/main/src/hydra_cli.py)